### DISCUSSION TITLE
Redirection behavior for disabled sites

### DISCUSSION THREAD

**Discussion by rtrudel** [craftcms/cms#14220] [Ideas] [answered]
I recently had to use the site status (enabled/disabled) and it didn't have the effect I expected. While I understand why it is this way by default, I would like to suggest better control over this:

A new general setting that allows deciding Craft's behavior when a disabled site is loaded by its base URL. The current behavior, when navigating to the URL of a disabled site, is to load the primary site, without redirection to the primary site's URL.

It would be good to be able to decide to redirect to the primary site's URL, or to force a 404 error, or make another redirection.

I would suggest something like 
```php
->redirectOnDisabledSite(null | true | false | String $redirect)
```

-  `null` (default) to keep the default behavior (loading the primary site without redirect)
- `true` to redirect to primary site URL (with HTTP code 3xx?)
- `false` to act same as template or page not found (with HTTP code 404 or any server error 5xx?) 
- `String` fully-qualified URL to redirect anywhere (with HTTP code 3xx?)

and a corresponding environment variable `REDIRECT_ON_DISABLED_SITE`.

This would allow for proper control over what we want to do with disabled sites.

Here is some Twig dirty hack I wrote to achieve something like this:

```twig
{% set REDIRECT_ON_DISABLED_SITE = getenv('REDIRECT_ON_DISABLED_SITE') %}
{% if not REDIRECT_ON_DISABLED_SITE == null and not ((craft.app.request.hostInfo|trim('/', 'right')~'/') == currentSite.baseUrl)  %}
    {% switch REDIRECT_ON_DISABLED_SITE %}
        {% case false %}
            {% exit 404 %} {# or another http code? #}
        {% case true %}
            {% redirect craft.app.sites.primarySite.baseUrl 302 %} {# or 301? #}
        {% default %}
            {% redirect REDIRECT_ON_DISABLED_SITE 302 %} {# or 301? #}
   {% endswitch %}
{% endif %}
```

**Comment by brandonkelly** [comment, marked as answer]
> The current behavior, when navigating to the URL of a disabled site, is to load the primary site, without redirection to the primary site's URL.

Yes and no… disabled sites aren’t loaded to begin with. They are completely excluded from front-end routing. So _yes_ you will probably end up seeing the primary site, but not because the disabled site choose to show it; because if Craft can’t match the incoming URL to any (non-disabled) site’s base URL, the primary site is loaded by default.

So a better way of thinking about this is: How should Craft handle _any_ incoming request where the URL doesn’t exactly match a site? We could potentially allow you to enable a “strict mode” that causes these requests to 404. And then that behavior would apply equally whether the request URL matches a disabled site, or no site at all.

**Reply by rtrudel** [reply]
I agree. A "strict mode" can be the proper way to deal with disabled sites, and all URLs which do not match any of the enabled sites. In a project, I had special case to deal with, and it made me think about using disabled site as a feature for the project.

My "edge case" was the following: I have a main site and some landing pages about promotions which are not always online. When a promo is active, the site owner can turn on a landing page about it, which resides on special subdomain. If the promotion is over and the landing page is turned off, I wanted a redirection to the regular product catalog. This only to avoid the promo site to be reachable while no promo is active, and not load the main site on the promo site url.

This very case can be handled inside a module, or in the main twig of the subdomain I guess. But I like your strict mode idea too.
